### PR TITLE
refactor: replace `peaks` method with `as.list` (issue #62)

### PR DESCRIPTION
**Pull request by jorainer**
- Replace `peaks` with `as.list` and `peaks<-` with `replaceList`.
- Update documentation and unit tests.

### CODE REVIEW

File: R/MsBackend.R [hunk 1]
```diff
@@ -125,6 +125,12 @@ NULL
 #'   spectrum. Returns an `integer` of length equal to the number of
 #'   spectra (with `NA_integer_` if not available).
 #'
+#' - `as.list` returns a `list` with the spectras' peak data. The length of
+#'   the list is equal to the number of spectra in `object`. Each element of
+#'   the list is a `matrix` with columns `mz` and `intensity`. For an empty
```
Comment (sgibb) [inline R/MsBackend.R]: Here you use:
> with columns `mz` and `intensity`

and for `replaceList` you use

> with columns `"mz"` and `"intensity"`

Please keep one of them. (I think the `"column"` version is easier to understand.)

File: R/MsBackend.R [hunk 2]
```diff
@@ -287,6 +283,12 @@ NULL
 #'   the number of spectra in `object`. `NA` are reported for MS1
 #'   spectra of if no precursor information is available.
 #'
+#' - `replaceList<-` replaces the peak data (m/z and intensity values) of the
+#'   backend. This method expects a `list` of `matrix` objects with columns
+#'   `"mz"` and `"intensity"` that has the same length than the number of
```
Comment (sgibb) [inline R/MsBackend.R]: See comment above.

Comment (sgibb) [inline R/MsBackend.R]:
```suggestion
#'   `"mz"` and `"intensity"` that has the same length as the number of
```

File: R/MsBackendHdf5Peaks.R [hunk 3]
```diff
@@ -157,7 +175,7 @@ setMethod("isEmpty", "MsBackendHdf5Peaks", function(x) {
 
 #' @rdname hidden_aliases
 setMethod("mz", "MsBackendHdf5Peaks", function(object) {
-    NumericList(lapply(peaks(object), "[", , 1), compress = FALSE)
+    NumericList(lapply(as.list(object), "[", , 1), compress = FALSE)
```
Comment (sgibb) [inline R/MsBackendHdf5Peaks.R]: This `list(lapply(as.list(` statement looks great now :wink:

File: R/MsBackendHdf5Peaks.R [hunk 4]
```diff
@@ -172,30 +190,16 @@ setReplaceMethod("mz", "MsBackendHdf5Peaks", function(object, value) {
...
-        unsplit(bpmapply(
-            FUN = .h5_read_peaks,
-            fls,
-            split(scanIndex(object), f),
-            object@modCount,
-            SIMPLIFY = FALSE, USE.NAMES = FALSE, BPPARAM = bpparam()),
-            f)
-    } else
-        .h5_read_peaks(fls, scanIndex(object), object@modCount)
+setMethod("peaksCount", "MsBackendHdf5Peaks", function(object) {
```
Comment (sgibb) [inline R/MsBackendHdf5Peaks.R]: Should this be `lengths` or so? Instead of **peaks**Count?

File: R/Spectra.R [hunk 5]
```diff
@@ -89,6 +89,11 @@ NULL
 #'   spectrum. Returns an `integer` of length equal to the number of
 #'   spectra (with `NA_integer_` if not available).
 #'
+#' - `as.list`: gets the *peaks* matrices for all spectra in `object`. The
+#'   function returns a [SimpleList()] of matrices, each `matrix` with columns
+#'   `mz` and `intensity` with the m/z and intensity values for all peaks of
```
Comment (sgibb) [inline R/Spectra.R]: See comment about `"mz"` vs `mz` and `"intensity"` vs `intensity` above.

File: R/MsBackend.R [hunk 6]
```diff
@@ -881,6 +874,13 @@ setMethod("precursorMz", "MsBackend", function(object) {
     stop("Not implemented for ", class(object), ".")
 })
 
+#' @exportMethod replaceList
+#'
+#' @rdname MsBackend
+setMethod("replaceList", "MsBackend", function(object, value, ...) {
```
Comment (sgibb) [inline R/MsBackend.R]: Could/Shouldn't this be a `setReplaceMethod` as well?

File: R/MsBackend.R (hunk 1 above)
Comment (jorainer) [reply R/MsBackend.R]: Yep. will change that. Thanks.

File: R/MsBackendHdf5Peaks.R (hunk 4 above)
Comment (jorainer) [reply R/MsBackendHdf5Peaks.R]: Sure! Yes! We said that. I will remove the `peaksCount` method.

File: R/MsBackend.R (hunk 6 above)
Comment (jorainer) [reply R/MsBackend.R]: I followed what they did in `S4Vectors`. The `replaceROW` and `replaceCOL` are **not** replace methods, they're just simple methods. I have no strong opinion here. If you want me to change I'll do.

File: R/MsBackendHdf5Peaks.R (hunk 4 above)
Comment (jorainer) [reply R/MsBackendHdf5Peaks.R]: Hm, only thing is that there's no generic `lengths`.

File: R/MsBackend.R (hunk 6 above)
Comment (sgibb) [reply R/MsBackend.R]: Me neither. Is there a performance drop? Or is `setReplaceMethod` internally just a synonym for `object <- setMethod`? (I don't have access to R at the moment.)

Comment (jorainer) [reply R/MsBackend.R]: Could one advantage be to have a dispatch on both `object` and `value`?

Comment (jorainer) [reply R/MsBackend.R]: Anyway, I'm going to change it to a replace method. Usage will then be:

```r
lst <- as.list(be)
replaceList(be) <- lst
```

I think that might be easier for users/developers to understand.

**Comment by jorainer** [comment]
Sorry for the many commits. What I did was:
- change `replaceList` to `replaceList<-`.
- fix/unify your comments on documentation.
- replace `peaksCount` by `lengths`.
- Fix unit tests for latest mzR (which caused Travis to fails).

File: R/MsBackend.R [hunk 7]
```diff
@@ -245,6 +253,10 @@ NULL
 #'
 #' - `length`: returns the number of spectra in the object.
 #'
+#' - `lengths`: gets the number of peaks (m/z-intensity values) per
+#'   spectrum.  Returns an `integer` vector (length equal to the
+#'   number of spectra). For empty spectra, `NA_integer_` is returned.
```
Comment (sgibb) [inline R/MsBackend.R]: Mh, does `lengths` really return `NA` for empty spectra? Not zero? Is `NA` a good choice here?

Comment (jorainer) [reply R/MsBackend.R]: Thanks for spotting this! The way it is implemented it would report `0` not `NA`. I'll fix the documentation.

**Pull request merged**